### DISCUSSION TITLE
Keycloak login page URL customization

### DISCUSSION THREAD

**Discussion by Eng-Fouad** [keycloak/keycloak#17690] [Q&A]
Is it possible to customize URL of the login page?

Something like:

> https://sso.example.com/realms/myrealm/protocol/openid-connect/auth

to:

> https://sso.example.com/login

**Comment by darius-m** [comment]
As far as I know that would not be possible since Keycloak fundamentally relies on realms to separate users and client configurations - e.g. the master realm is for administration purposes only, while other realms serve clients - which makes the `/realms/myrealm` component mandatory. The `openid-connect/auth` is used to identify the authentication protocol (either OIDC or SAML2) and how the request is parsed and parameter handling depends on it.

**Reply by Eng-Fouad** [reply]
It would be possible if we are able to register alias path that resolves at server side.

What about url-rewrite at reverse-proxy layer? Would that work?

**Reply by darius-m** [reply]
I'm not sure, I have not tried it myself. It may be possible to do it through the reverse proxy, since it terminates the client connection and sends a (possibly modified) request to the backend, which can in turn be directed to a different endpoint. However, I assume that it would also limit your access to a single realm and protocol. It is also likely possible to change the endpoint format using an SPI, which would basically create additional endpoints (like the one you have mentioned - `/login`).

Would you mind sharing why accessing the authentication endpoint in this manner would be required? Is it simply cosmetic, or is there a different requirement?

**Reply by gluggi** [reply]
Hi, I have similar requirements and challanges to get it done as I'm not able to make a working reverse proxy setup.
1.) it is somehow cosmetics and it would allow to have a URL e.g. for
* realm Prod1 https://prod1.com/auth 
and for realm 
* Prod2 realm https://prod2.com/auth 
with all the openid-connect possibilities

2.) it is a security feature as you cannot easily butforce check what other realms are setup and what kind of openid-connect implementation (here keycloak) is used.

using the Frontend Url in the realm configuration is working partly.. except it allways is adding the realm/MYREALM/ part to it and ofcaurse all the js and css stuff which is loaded afterwards.

So this would be a realy nice feature.. at least for me
Is this request for setup realy so special and is everyone exposing all realms to the outside world?

Best 
Martin

**Reply by StefanLobbenmeierObjego** [reply]
> Is it simply cosmetic, or is there a different requirement?

I have the same request, in our case its mostly cosmetics but also we are migrating from direct grant flow to the regular browser authentication flow, but would like to avoid disruptions to our users. Ideally the URL stays mostly the same so that users would only notice a slight change in the design and be able to trust that they are still providing username + password to us and not some other site.

**Comment by vplemyannik** [comment]
@Eng-Fouad you can use direct grant flow, and implement your own login page with url you like
(But this approach is considered less secured)

**Reply by gluggi** [reply]
Thanks for the reply, but it cannot be the solution I need to write my own login screens incl. password reminder ... and  for social indentity provider . This makes the use of this product useless and I can implement it directly in my backend.

On the one side it is cosmetics.. on the other side it shoudl bot be possible that someone with knowledge can just change the url to spoof  other existing realms. So it is also an security erquirement.

Are we realy the only once with the wish to hide https://bla.com/realms/deltra/account/ behind something like 
https://bla.com/account/
or even 
https://auth.bla.com/
?

**Reply by allenjoseph29** [reply]
Nope. Even our team is also looking for a similar solution. @gluggi did you guys figure it out?

**Reply by gluggi** [reply]
Hi, eventually we moved away from keycloak. So the answer is no.

Cheers

**Reply by nightvision04** [reply]
@gluggi @allenjoseph29 What did you guys end up going with? 

Our team is strongly considering it too, but I'm a bit disappointed with the lack of customization.
